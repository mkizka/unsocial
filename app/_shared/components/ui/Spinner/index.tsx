import { cn } from "@/_shared/utils/cn";

type Props = {
  className?: string;
};

// https://zenn.dev/catnose99/articles/19a05103ab9ec7
export function Spinner({ className }: Props) {
  return (
    <div
      className={cn(
        "flex h-full w-full items-center justify-center",
        className,
      )}
    >
      <div className="aspect-square h-4/5 animate-spin rounded-full border-2 border-t-transparent"></div>
    </div>
  );
}
