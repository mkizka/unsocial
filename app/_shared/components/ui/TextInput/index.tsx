import { cn } from "@/_shared/utils/cn";

// https://zenn.dev/andynuma/articles/c7f6d6587c116d
type TextInputProps<T extends React.ElementType> = {
  as?: T;
};

type Props<T extends React.ElementType> = TextInputProps<T> &
  Omit<React.ComponentProps<T>, keyof TextInputProps<T>>;

export function TextInput<T extends "input" | "textarea" = "input">({
  className,
  as,
  ...props
}: Props<T>) {
  const TagName = as || "input";
  return (
    // @ts-expect-error
    <TagName
      className={cn(
        {
          "form-input": TagName === "input",
          "form-textarea": TagName === "textarea",
        },
        "block w-full rounded border-primary-dark bg-primary-light",
        "focus:border-primary-dark focus:ring focus:ring-primary-dark focus:ring-opacity-50",
        className,
      )}
      {...props}
    />
  );
}
