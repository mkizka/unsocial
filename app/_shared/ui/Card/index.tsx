import { cn } from "@/_shared/utils/cn";

// https://zenn.dev/andynuma/articles/c7f6d6587c116d
type CardProps<T extends React.ElementType> = {
  as?: T;
  children: React.ReactNode;
};

type Props<T extends React.ElementType> = CardProps<T> &
  Omit<React.ComponentProps<T>, keyof CardProps<T>>;

export function Card<T extends React.ElementType = "div">({
  children,
  className,
  as,
  ...props
}: Props<T>) {
  const TagName = as || "div";
  return (
    <TagName
      className={cn(
        "rounded-lg border bg-card p-4 text-card-foreground shadow-sm",
        className,
      )}
      {...props}
    >
      {children}
    </TagName>
  );
}
