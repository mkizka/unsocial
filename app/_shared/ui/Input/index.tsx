import { type ComponentPropsWithoutRef, forwardRef } from "react";

import { cn } from "@/_shared/utils/cn";

export const Input = forwardRef<
  HTMLInputElement,
  ComponentPropsWithoutRef<"input">
>(({ className, ...props }, ref) => {
  return (
    <input
      className={cn(
        "flex h-10 w-full rounded-md border border-input bg-input-inner px-3 py-2 text-sm ring-offset-background",
        "file:border-0 file:bg-transparent file:text-sm file:font-medium",
        "placeholder:text-muted-foreground",
        "focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring",
        "disabled:cursor-not-allowed disabled:opacity-50",
        className,
      )}
      ref={ref}
      {...props}
    />
  );
});
Input.displayName = "Input";
