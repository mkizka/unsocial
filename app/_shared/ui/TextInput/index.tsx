import { cn } from "@/_shared/utils/cn";

// https://zenn.dev/andynuma/articles/c7f6d6587c116d
type TextInputProps<T extends React.ElementType> = {
  as?: T;
};

type Props<T extends React.ElementType> = TextInputProps<T> &
  Omit<React.ComponentProps<T>, keyof TextInputProps<T>>;

export function TextInput<T extends "input" | "textarea" = "input">({
  className,
  as,
  ...props
}: Props<T>) {
  const TagName = as || "input";
  return (
    // @ts-expect-error
    <TagName
      className={cn(
        {
          "form-input": TagName === "input",
          "form-textarea": TagName === "textarea",
        },
        "bg-primary-light block w-full rounded",
        "focus:ring-primary-dark focus:focus:ring focus:ring-opacity-50",
        className,
      )}
      {...props}
    />
  );
}
