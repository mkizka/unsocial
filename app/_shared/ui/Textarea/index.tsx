import type { ComponentPropsWithoutRef } from "react";

import { cn } from "@/_shared/utils/cn";

export function Textarea({
  className,
  ...props
}: ComponentPropsWithoutRef<"textarea">) {
  return (
    <textarea
      className={cn(
        "flex min-h-[80px] w-full rounded-md border border-input bg-input-inner px-3 py-2 text-sm ring-offset-background",
        "placeholder:text-muted-foreground",
        "focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring",
        "disabled:cursor-not-allowed disabled:opacity-50",
        className,
      )}
      {...props}
    />
  );
}
