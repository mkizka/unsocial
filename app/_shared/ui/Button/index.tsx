"use client";
import { cva, type VariantProps } from "class-variance-authority";
import { forwardRef } from "react";
import { useFormStatus } from "react-dom";

import { Spinner } from "@/_shared/ui/Spinner";
import { cn } from "@/_shared/utils/cn";

const buttonVariants = cva(
  cn(
    "relative flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors",
    "disabled:pointer-events-none disabled:opacity-50",
  ),
  {
    variants: {
      variant: {
        default:
          "bg-primary text-primary-foreground shadow hover:bg-primary/90",
        destructive:
          "bg-destructive text-destructive-foreground shadow-sm hover:bg-destructive/90",
        outline:
          "border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground",
        secondary:
          "bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80",
        ghost: "hover:bg-secondary",
        link: "text-primary underline-offset-4 hover:underline",
      },
      size: {
        default: "h-9 px-4 py-2",
        sm: "h-8 rounded-md px-3 text-xs",
        lg: "h-10 rounded-md px-8",
        icon: "size-8",
      },
    },
    defaultVariants: {
      variant: "default",
      size: "default",
    },
  },
);

interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  asChild?: boolean;
  loading?: boolean;
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      className,
      variant,
      size,
      children,
      asChild = false,
      loading = false,
      ...props
    },
    ref,
  ) => {
    const status = useFormStatus();
    const isLoading = status.pending || loading;
    return (
      <button
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...props}
      >
        {isLoading && <Spinner className="absolute h-1/2" />}
        <span
          className={cn(
            "flex justify-center items-center w-full h-full gap-1",
            {
              "opacity-0": isLoading,
            },
          )}
        >
          {children}
        </span>
      </button>
    );
  },
);
Button.displayName = "Button";
