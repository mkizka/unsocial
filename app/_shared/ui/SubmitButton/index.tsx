import type { ComponentProps, ReactNode } from "react";
import { useFormStatus } from "react-dom";

import { Spinner } from "@/_shared/ui/Spinner";
import { cn } from "@/_shared/utils/cn";

type Props = ComponentProps<"button"> & {
  children?: ReactNode;
};

export function SubmitButton({ children, className, ...props }: Props) {
  const status = useFormStatus();
  return (
    <button
      className={cn(
        "relative flex items-center justify-center",
        "rounded bg-secondary px-3 py-1.5 shadow",
        "hover:bg-secondary-dark disabled:bg-secondary-dark",
        className,
      )}
      type="submit"
      data-testid="submit-button"
      disabled={status.pending}
      {...props}
    >
      {status.pending && <Spinner className="absolute h-4/5" />}
      <span
        className={cn({
          "opacity-0": status.pending,
        })}
      >
        {children}
      </span>
    </button>
  );
}
