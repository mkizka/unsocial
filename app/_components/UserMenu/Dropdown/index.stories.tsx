import { expect } from "@storybook/jest";
import type { Meta, StoryObj } from "@storybook/react";
import { userEvent, within } from "@storybook/testing-library";
import type { ComponentProps } from "react";

import { Card } from "@/_shared/components/ui/Card";

import { Dropdown } from ".";

function Component(args: ComponentProps<typeof Dropdown>) {
  return (
    <Card className="flex w-64 justify-end">
      <Dropdown {...args} />
    </Card>
  );
}
const meta: Meta<typeof Dropdown> = {
  component: Component,
  args: {
    iconUrl: "https://via.placeholder.com/36x36",
    iconAlt: "アイコン",
  },
};

export default meta;

type Story = StoryObj<typeof Dropdown>;

export const Default: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    await userEvent.click(canvas.getByTestId("user-menu__button"));
    expect(canvas.queryByTestId("user-menu__dropdown")).toBeInTheDocument();
    await userEvent.click(canvas.getByTestId("user-menu__backdrop"));
    expect(canvas.queryByTestId("user-menu__dropdown")).not.toBeInTheDocument();
  },
};
