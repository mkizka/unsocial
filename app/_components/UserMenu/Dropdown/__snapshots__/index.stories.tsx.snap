// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_components/UserMenu/Dropdown Default play-test 1`] = `
<div class="rounded bg-primary-light p-4 shadow flex w-64 justify-end">
  <div class="relative size-9"
       data-testid="user-menu"
  >
    <button class="hover:opacity-70"
            data-testid="user-menu__button"
    >
      <img class="rounded-full"
           src="https://via.placeholder.com/36x36"
           alt="アイコン"
           width="36"
           height="36"
      >
    </button>
  </div>
</div>
`;
